app/models/population.rb

```diff
@@ -0,0 +1,56 @@
+class Population < ApplicationRecord
+  belongs_to :location
+
+  validates :date, presence: true
+  validates :operational_capacity, presence: true, numericality: { greater_than: 0 }
+  validates :usable_capacity, presence: true, numericality: { greater_than: 0 }
+  validates :unlock, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :bedwatch, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :overnights_in, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :overnights_out, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :out_of_area_courts, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :discharges, presence: true, numericality: { greater_than_or_equal_to: 0 }
+
+  def moves_from
+    location.moves_from.prison_transfer.not_cancelled.where(date: date)
+  end
+
+  def moves_to
+    location.moves_to.prison_transfer.not_cancelled.where(date: date)
+  end
+
+  def free_spaces
+    usable_capacity - unlock - bedwatch - overnights_in - moves_to.count + overnights_out + out_of_area_courts + discharges + moves_from.count
+  end
+
+  def self.free_spaces_date_range(locations, date_range)
+    {}.tap do |locations_hash|
+      date_range.each do |date|
+        free_spaces_for_date(locations, date).each do |row|
+          location_id, population_id, free_spaces = *row
+
+          free_space_details = { id: population_id, free_spaces: free_spaces } if population_id.present?
+
+          locations_hash[location_id] = [] unless locations_hash.key?(location_id)
+          locations_hash[location_id] << free_space_details
+        end
+      end
+    end
+  end
+
+  def self.free_spaces_for_date(locations, date)
+    locations
+      # Join with matching populations for location and given date (if any). Can't use a where clause as there may not be a capacity record
```

Comment:
```suggestion
      # Join with matching populations for location and given date (if any). Can't use a where clause as there may not be a population record
```

Reply:
Maybe this model could be `LocationCapacityInformation`? I find `Population` less intuitive? Essentially these are descriptive statistics about a location on a given day? Are the locations restricted to prisons or do we report information about other types of location?

Reply:
On the front end the adopted terminology is population so I've aligned with that just to avoid the impedance mismatch when discussing this. My initial choice during the spike was "capacity" but I've changed to "population" now.

Reply:
😢  okay